stm32: refcount for backup domain access protection management

etienne-lms opened the pull request:
Add STM32 SoC functions `stm32_backup_domain_enable_access()` and `stm32_backup_domain_disable_access()` to enable/disable access to backup domain resources using a reference counter.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/90942

etienne-lms committed:
soc: st: stm32: reference counting for backup domain accesses

Add SoC functions to enable/disable STM32 backup Domain access
and use a reference counter to track requests. These helper functions
may be called from a interrupt context. On domain access enable, the
function loops until written bit is set however this is expected to be
effective after very few clock cycles and seems not even required
(not mentioned in any SoC documentation). The loop is preserved as
used in previous implementation.

Among all supported STM32 SoCs, only STM32C0 and STM32WB0 series do not
implement this mechanism hence add option CONFIG_STM32_BACKUP_PROTECTION
that is enabled for all SoC series but these 2.

Fixes issue 90942.

Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>

etienne-lms committed:
soc: st: stm32: reference counter for Backup SRAM accesses

Use recently added SoCs functions to request access to SoC backup
domain resources These function use a reference counter to track these
request and ensure the resources are accessible as long as at least a
consumer requires access.

Fixes issue 90942.

Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>

etienne-lms committed:
drivers: bbram: stm32: reference counter for Backup Registers accesses

Use recently added SoCs functions to request access to SoC backup
domain resources These function use a reference counter to track these
request and ensure the resources are accessible as long as at least a
consumer requires access.

Fixes issue 90942.

Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>

etienne-lms committed:
drivers: clock_control: stm32: refcounter for Backup domain accesses

Use recently added SoCs functions to request access to SoC backup
domain resources These function use a reference counter to track these
request and ensure the resources are accessible as long as at least a
consumer requires access.

By the way, correct stm32_hsem.h header file inclusion that requires
brackets (<>) delimiters, not double quotes, as per convention on header
location.

Fixes issue 90942.

Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>

etienne-lms committed:
drivers: rtc: stm32: refcounter for Backup domain accesses

Use recently added SoCs functions to request access to SoC backup
domain resources These function use a reference counter to track these
request and ensure the resources are accessible as long as at least a
consumer requires access.

By the way, move inclusion of stm32_hsem.h header file after inclusion
of the LL header files and Zephyr header standard files.

Fixes issue 90942.

Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>

etienne-lms committed:
drivers: counter: stm32: refcounter for Backup domain accesses

Use recently added SoCs functions to request access to SoC backup
domain resources These function use a reference counter to track these
request and ensure the resources are accessible as long as at least a
consumer requires access. When we're done accessing RTC registers and
other protected resource, release the access refcount.

By the way, simplify rtc_stm32_read() when COUNTER_NO_DATE is defined
that used a useless extra local variable.

By the way, correct stm32_hsem.h header file inclusion that requires
brackets (<>) delimiters, not double quotes, as per convention header
location.

Fixes issue 90942.

Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>

etienne-lms committed:
drivers: clock_control: stm32: release Backup Domain access refcount

Fix STM32 WBA and H7 clock drivers to release the reference counter
added to access LSE configuration controllers once the clock is
configured. Keeping such an unbalanced access request is no more
needed since SoC functions manage reference counting of access requests.

Fixes issue 90942.

Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>

mathieuchopstm commented on soc/st/stm32/common/stm32_backup_domain.h:
```diff
@@ -0,0 +1,69 @@
+#endif
+
+#ifdef CONFIG_STM32_BACKUP_PROTECTION
+struct stm32_backup_domain_access {
+	struct k_spinlock lock;
+	size_t refcount;
+};
+
+extern struct stm32_backup_domain_access stm32_backup_domain_access;
```
Related: if we move implementation to `.c`, this can be removed from the header and kept private.

mathieuchopstm reviewed (commented):
Since only 2 series are excluded, I'd lean more towards `default y if !SERIES_W_NO_DBP` rather than select everywhere, but that's a detail..

etienne-lms reviewed (commented):
> Somewhat unrelated to the core of the PR, but putting a link to an issue in the commit message pollutes the issue page on GitHub.

Maybe changing the commit message line could prevent that while keeping the reference, if it helps...
```patch
-Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/90942
+Fixes issue 90942

etienne-lms commented:
> Since only 2 series are excluded, I'd lean more towards default y if !SERIES_W_NO_DBP rather than select everywhere, but that's a detail..

Addressed in the 1st commit of the series.

mathieuchopstm commented on soc/st/stm32/common/stm32_backup_domain.c:
Wakeup pins are not supported yet on all series (c.f. #81119), doing this will very much likely break everything. And some power stuff is SoC-specific, not global.

etienne-lms commented:
I missed to consider MP257F specific case. Addressed now. 
I've also fixed issues reported by checkpatch.

gautierg-st commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -240,11 +241,7 @@ tick_t rtc_stm32_read(const struct device *dev)
-	defined(PWR_DBPCR_DBP) || defined(PWR_DBPR_DBP)
-	LL_PWR_EnableBkUpAccess();
-#endif /* PWR_CR_DBP || PWR_CR1_DBP || PWR_DBPR_DBP */
+	stm32_backup_domain_enable_access();
```
It seems like the refcount won't be balanced at all here. We'll keep incrementing it each time we'll read, and never decrementing it.
Since it is already enabled at init, I think it isn't necessary to keep it here.

etienne-lms replied:
Indeed, thanks!

etienne-lms replied:
Addressed

mathieuchopstm commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -18,6 +18,7 @@
 #include <zephyr/sys/util.h>
 #include <zephyr/kernel.h>
 #include <soc.h>
+#include <stm32_backup_domain.h>
```
Nit: I would group this below with `stm32_hsem.h` - file could even be included using `"stm32_backup_domain"`, I think

Ditto in all files.

etienne-lms replied:
I'd prefer to keep `#include "foo.h"` for local header files.

mathieuchopstm reviewed (commented):
Unrelated to PR, but this section in `counter_ll_stm32_rtc.c` is a bit odd:
![image](https://github.com/user-attachments/assets/b9e2583d-1b63-47b9-bbdd-0a7774bbaf5f)

LGTM otherwise with a very minor nit:

mathieuchopstm commented on soc/st/stm32/common/stm32_backup_domain.c:
```diff
@@ -0,0 +1,60 @@
+#define IS_ENABLED_BKUP_ACCESS	LL_PWR_IsEnabledBkUpAccess
+#endif
+
+struct stm32_backup_domain_access {
+	struct k_spinlock lock;
+	size_t refcount;
+};
+
+static struct stm32_backup_domain_access stm32_backup_domain_access;
```
CI is not happy because the object has the same name as the structure.

Either change the object's name, or drop the structure entirely - this is private now, so using a bare `k_spinlock`+`size_t` is fine.

erwango commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -568,16 +563,15 @@ static int rtc_stm32_init(const struct device *dev)
 				    (clock_control_subsys_t) &cfg->pclken[1],
 				    NULL) != 0) {
 		LOG_ERR("clock configure failed\n");
+		stm32_backup_domain_disable_access();
 		return -EIO;
```
[Non-blocking] While we're at, what about a goto to the functions exit ?

erwango reviewed (dismissed):
Neat, thanks!

etienne-lms commented:
I've added the commit from https://github.com/zephyrproject-rtos/zephyr/pull/92075 (now closed) as that P-R depended on this current P-R.

etienne-lms commented:
Updated regarding recently merged https://github.com/zephyrproject-rtos/hal_stm32/pull/290 dependency.

etienne-lms committed:
soc: st: stm32wbax: refcount backup domain accesses requests

Add LINKLAYER_PLAT_EnableBackupDomainAccess()  and
LINKLAYER_PLAT_DisableBackupDomainAccess() to use Zephyr resources
that use a reference counter for access requests, for enabling
and disabling access the BackupDomain resources.

Bump hal_stm32 module to the revision integrating related stm32wba
BLE updates.

Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>

etienne-lms commented:
Rebased to address the DNM label related to the hal_stm32 module revision update.
It seems still not resolved. I don't understand where the problem is. 
Any help is welcome!

erwango commented:
Removing the flags manually as somehow they seem sticky.

mathieuchopstm commented:
> Removing the flags manually as somehow they seem sticky.

Not sure removing `Binary Blobs added` was [necessary](https://github.com/zephyrproject-rtos/zephyr/blob/3370e524ca43c9a542c06e9bf8c716c68397cb15/scripts/ci/do_not_merge.py#L12).
